Elektra Web 1.3

omnidan opened the pull request:
@markus2330 I finished (almost) all tasks from #1865 and tested the app with Firefox ESR and the real-world config.

I also fixed the peer-dependency warnings and warnings about missing fields in `package.json`s :grin: The cmake integration now runs `npm prune`, which removes dev/build dependencies after building (reduces size of `node_modules`, the total build size of elektra-web is 288MB for me now).

The only tasks missing now are reworking the documentation, which I will still do before this PR gets merged (probably tomorrow).

Please feel free to test & review the new version already.

BTW: drag & drop works for me in Firefox ESR

---

I moved the documentation tasks here, so the other issue (#1865) can be closed now:

## ToDo

- [x] kdb metakey newline issue (debian?)
- [x] Please also make sure to disable the tests if dredd is not available.
- [x] huge hint in the "create new key"
- [x] order of array keys which partly have subkeys (the one with subkeys are before the keys without subkeys) is irritating

## ToDo (documentation / api blueprints)

- [x] elektrad
- [x] webd
- [x] Response 204 not documented?
- [x] how can you set metadata?
- [x] delete recursive?
- [x] Some kind of automatic testing of elektrad and the web-ui would be very much appreciated.

omnidan commented:
I am using `dredd` to automatically run api tests from the blueprints. I already updated the blueprint for elektrad (still working on webd). You can run the tests by starting elektrad and then executing `npm test` in the `elektrad` folder:

```
> npm test

> @elektra-web/elektrad@1.3.0 test /Users/dan/Development/git/hub/omnidan/libelektra/src/tools/web/elektrad
> dredd ../../../../doc/api_blueprints/elektrad.apib http://localhost:33333

info: Beginning Dredd testing...
pass: GET (200) /version duration: 95ms
pass: GET (200) /kdb/user/hello duration: 44ms
pass: GET (200) /kdb/user/doesnotexist duration: 42ms
pass: PUT (204) /kdb/user/hello duration: 46ms
pass: DELETE (204) /kdb/user/hello duration: 28ms
pass: POST (204) /kdbMeta/user/hello duration: 35ms
pass: DELETE (204) /kdbMeta/user/hello duration: 33ms
pass: POST (204) /kdbMv/user/hello duration: 34ms
pass: POST (204) /kdbCp/user/hello duration: 25ms
complete: 9 passing, 0 failing, 0 errors, 0 skipped, 9 total
complete: Tests took 393ms
```

markus2330 commented:
`npm start` of elektrad failed again (after successfull `npm install`):

```
> @elektra-web/elektrad@1.3.0 start /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad
> npm run build && npm run start:prod


> @elektra-web/elektrad@1.3.0 build /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad
> cross-env NODE_ENV=production babel -d build/ src/

src/app.js -> build/app.js
src/config.js -> build/config.js
src/index.js -> build/index.js
src/log.js -> build/log.js
src/routes/index.js -> build/routes/index.js
src/routes/utils.js -> build/routes/utils.js
src/versions.js -> build/versions.js

> @elektra-web/elektrad@1.3.0 start:prod /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad
> cross-env NODE_ENV=production DEBUG=elektrad:*,-*debug,-*log node build/

  elektrad:info @elektra-web/elektrad v1.3.0 starting +0ms
  elektrad:error you are running an old libelektra version, which is not supported +23ms
  elektrad:error please upgrade to libelektra 0.8.23 or higher +1ms

npm ERR! Linux 4.12.0-0.bpo.2-amd64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "run" "start:prod"
npm ERR! node v6.13.1
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! @elektra-web/elektrad@1.3.0 start:prod: `cross-env NODE_ENV=production DEBUG=elektrad:*,-*debug,-*log node build/`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @elektra-web/elektrad@1.3.0 start:prod script 'cross-env NODE_ENV=production DEBUG=elektrad:*,-*debug,-*log node build/'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the @elektra-web/elektrad package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     cross-env NODE_ENV=production DEBUG=elektrad:*,-*debug,-*log node build/
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs @elektra-web/elektrad
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls @elektra-web/elektrad
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad/npm-debug.log

npm ERR! Linux 4.12.0-0.bpo.2-amd64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! node v6.13.1
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! @elektra-web/elektrad@1.3.0 start: `npm run build && npm run start:prod`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @elektra-web/elektrad@1.3.0 start script 'npm run build && npm run start:prod'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the @elektra-web/elektrad package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run build && npm run start:prod
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs @elektra-web/elektrad
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls @elektra-web/elektrad
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
```

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'start' ]
2 info using npm@3.10.10
3 info using node@v6.13.1
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle @elektra-web/elektrad@1.3.0~prestart: @elektra-web/elektrad@1.3.0
6 silly lifecycle @elektra-web/elektrad@1.3.0~prestart: no script for prestart, continuing
7 info lifecycle @elektra-web/elektrad@1.3.0~start: @elektra-web/elektrad@1.3.0
8 verbose lifecycle @elektra-web/elektrad@1.3.0~start: unsafe-perm in lifecycle true
9 verbose lifecycle @elektra-web/elektrad@1.3.0~start: PATH: /usr/lib/node_modules/npm/bin/node-gyp-bin:/home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad/node_modules/.bin:/usr/lib/git-core:/usr/local/bin:/usr/games:/usr/bin:/bin:/home/markus/bin:/opt/idea/bin:.
10 verbose lifecycle @elektra-web/elektrad@1.3.0~start: CWD: /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad
11 silly lifecycle @elektra-web/elektrad@1.3.0~start: Args: [ '-c', 'npm run build && npm run start:prod' ]
12 silly lifecycle @elektra-web/elektrad@1.3.0~start: Returned: code: 1  signal: null
13 info lifecycle @elektra-web/elektrad@1.3.0~start: Failed to exec start script
14 verbose stack Error: @elektra-web/elektrad@1.3.0 start: `npm run build && npm run start:prod`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/lib/utils/lifecycle.js:255:16)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at EventEmitter.emit (events.js:191:7)
14 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/lib/utils/spawn.js:40:14)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at ChildProcess.emit (events.js:191:7)
14 verbose stack     at maybeClose (internal/child_process.js:920:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:230:5)
15 verbose pkgid @elektra-web/elektrad@1.3.0
16 verbose cwd /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad
17 error Linux 4.12.0-0.bpo.2-amd64
18 error argv "/usr/bin/node" "/usr/bin/npm" "start"
19 error node v6.13.1
20 error npm  v3.10.10
21 error code ELIFECYCLE
22 error @elektra-web/elektrad@1.3.0 start: `npm run build && npm run start:prod`
22 error Exit status 1
23 error Failed at the @elektra-web/elektrad@1.3.0 start script 'npm run build && npm run start:prod'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the @elektra-web/elektrad package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     npm run build && npm run start:prod
23 error You can get information on how to open an issue for this project with:
23 error     npm bugs @elektra-web/elektrad
23 error Or if that isn't available, you can get their info via:
23 error     npm owner ls @elektra-web/elektrad
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

omnidan commented:
@markus2330 neither elektrad nor webd will start with an elektra version that is lower than 0.8.23, please upgrade elektra or build via the cmake integration.

>   elektrad:error you are running an old libelektra version, which is not supported +23ms
>   elektrad:error please upgrade to libelektra 0.8.23 or higher +1ms

markus2330 commented:
Thank you for the tip. There is so much output that this is not so easy to see (if you are not used to development with npm).

Now I get "Not Found" when trying to connect to webd. Webd outputs in the console:

```
> @elektra-web/webd@1.3.0 start /home/markus/Projekte/Elektra/libelektra/src/tools/web/webd
> npm run build && npm run start:prod


> @elektra-web/webd@1.3.0 build /home/markus/Projekte/Elektra/libelektra/src/tools/web/webd
> cross-env NODE_ENV=production babel -d build/ src/

src/app.js -> build/app.js
src/config.js -> build/config.js
src/connector.js -> build/connector.js
src/db/clusters.js -> build/db/clusters.js
src/db/config.js -> build/db/config.js
src/db/index.js -> build/db/index.js
src/db/instances.js -> build/db/instances.js
src/db/utils.js -> build/db/utils.js
src/index.js -> build/index.js
src/log.js -> build/log.js
src/routes/clusters.js -> build/routes/clusters.js
src/routes/index.js -> build/routes/index.js
src/routes/instances.js -> build/routes/instances.js
src/routes/utils.js -> build/routes/utils.js
src/serveClient.js -> build/serveClient.js
src/versions.js -> build/versions.js

> @elektra-web/webd@1.3.0 start:prod /home/markus/Projekte/Elektra/libelektra/src/tools/web/webd
> cross-env NODE_ENV=production DEBUG=webd:*,-*debug,-*log node build/

  webd:info @elektra-web/webd v1.3.0 starting +0ms
  webd:info |- using default kdb command +41ms
  webd:info |- versions: { api: 1, elektra: { version: '0.8.23', major: 0, minor: 8, micro: 23 } } +1ms
  webd:client:info prod mode: serving client static assets +7ms
  webd:info `-> running on http://localhost:33334 +3ms
Error: ENOENT: no such file or directory, stat '/home/markus/Projekte/Elektra/libelektra/src/tools/web/client/build/index.html'
    at Error (native)
Error: ENOENT: no such file or directory, stat '/home/markus/Projekte/Elektra/libelektra/src/tools/web/client/build/index.html'
    at Error (native)
Error: ENOENT: no such file or directory, stat '/home/markus/Projekte/Elektra/libelektra/src/tools/web/client/build/index.html'
    at Error (native)
```

And yes, there is no "build" directory in the client. When trying to start the client (like it says in the top-level README.md) it fails:

```
> @elektra-web/client@1.3.0 start /home/markus/Projekte/Elektra/libelektra/src/tools/web/client
> npm run build && cd ../webd && npm run build && npm run start:prod


> @elektra-web/client@1.3.0 build /home/markus/Projekte/Elektra/libelektra/src/tools/web/client
> react-scripts build

Creating an optimized production build...
ls
Compiled successfully.

File sizes after gzip:

  163.36 KB  build/static/js/main.0562c4f3.js
  1.06 KB    build/static/css/main.ac1d4dbe.css

The project was built assuming it is hosted at the server root.
To override this, specify the homepage in your package.json.
For example, add this to build it for GitHub Pages:

  "homepage" : "http://myname.github.io/myapp",

The build folder is ready to be deployed.
You may serve it with a static server:

  npm install -g serve
  serve -s build


> @elektra-web/webd@1.3.0 build /home/markus/Projekte/Elektra/libelektra/src/tools/web/webd
> cross-env NODE_ENV=production babel -d build/ src/

src/app.js -> build/app.js
src/config.js -> build/config.js
src/connector.js -> build/connector.js
src/db/clusters.js -> build/db/clusters.js
src/db/config.js -> build/db/config.js
src/db/index.js -> build/db/index.js
src/db/instances.js -> build/db/instances.js
src/db/utils.js -> build/db/utils.js
src/index.js -> build/index.js
src/log.js -> build/log.js
src/routes/clusters.js -> build/routes/clusters.js
src/routes/index.js -> build/routes/index.js
src/routes/instances.js -> build/routes/instances.js
src/routes/utils.js -> build/routes/utils.js
src/serveClient.js -> build/serveClient.js
src/versions.js -> build/versions.js

> @elektra-web/webd@1.3.0 start:prod /home/markus/Projekte/Elektra/libelektra/src/tools/web/webd
> cross-env NODE_ENV=production DEBUG=webd:*,-*debug,-*log node build/

  webd:info @elektra-web/webd v1.3.0 starting +0ms
  webd:info |- using default kdb command +43ms
  webd:info |- versions: { api: 1, elektra: { version: '0.8.23', major: 0, minor: 8, micro: 23 } } +1ms
  webd:client:info prod mode: serving client static assets +7ms
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE :::33334
    at Object.exports._errnoException (util.js:1020:11)
    at exports._exceptionWithHostPort (util.js:1043:20)
    at Server._listen2 (net.js:1277:14)
    at listen (net.js:1313:10)
    at Server.listen (net.js:1409:5)
    at EventEmitter.listen (/home/markus/Projekte/Elektra/libelektra/src/tools/web/webd/node_modules/express/lib/application.js:618:24)
    at initApp (/home/markus/Projekte/Elektra/libelektra/src/tools/web/webd/build/app.js:52:7)
    at /home/markus/Projekte/Elektra/libelektra/src/tools/web/webd/build/index.js:61:25
    at process._tickCallback (internal/process/next_tick.js:109:7)

npm ERR! Linux 4.12.0-0.bpo.2-amd64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "run" "start:prod"
npm ERR! node v6.13.1
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! @elektra-web/webd@1.3.0 start:prod: `cross-env NODE_ENV=production DEBUG=webd:*,-*debug,-*log node build/`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @elektra-web/webd@1.3.0 start:prod script 'cross-env NODE_ENV=production DEBUG=webd:*,-*debug,-*log node build/'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the @elektra-web/webd package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     cross-env NODE_ENV=production DEBUG=webd:*,-*debug,-*log node build/
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs @elektra-web/webd
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls @elektra-web/webd
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/markus/Projekte/Elektra/libelektra/src/tools/web/webd/npm-debug.log

npm ERR! Linux 4.12.0-0.bpo.2-amd64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! node v6.13.1
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! @elektra-web/client@1.3.0 start: `npm run build && cd ../webd && npm run build && npm run start:prod`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @elektra-web/client@1.3.0 start script 'npm run build && cd ../webd && npm run build && npm run start:prod'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the @elektra-web/client package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run build && cd ../webd && npm run build && npm run start:prod
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs @elektra-web/client
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls @elektra-web/client
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/markus/Projekte/Elektra/libelektra/src/tools/web/client/npm-debug.log
```

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'start' ]
2 info using npm@3.10.10
3 info using node@v6.13.1
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle @elektra-web/client@1.3.0~prestart: @elektra-web/client@1.3.0
6 silly lifecycle @elektra-web/client@1.3.0~prestart: no script for prestart, continuing
7 info lifecycle @elektra-web/client@1.3.0~start: @elektra-web/client@1.3.0
8 verbose lifecycle @elektra-web/client@1.3.0~start: unsafe-perm in lifecycle true
9 verbose lifecycle @elektra-web/client@1.3.0~start: PATH: /usr/lib/node_modules/npm/bin/node-gyp-bin:/home/markus/Projekte/Elektra/libelektra/src/tools/web/client/node_modules/.bin:/usr/lib/git-core:/usr/local/bin:/usr/games:/usr/bin:/bin:/home/markus/bin:/opt/idea/bin:.
10 verbose lifecycle @elektra-web/client@1.3.0~start: CWD: /home/markus/Projekte/Elektra/libelektra/src/tools/web/client
11 silly lifecycle @elektra-web/client@1.3.0~start: Args: [ '-c',
11 silly lifecycle   'npm run build && cd ../webd && npm run build && npm run start:prod' ]
12 silly lifecycle @elektra-web/client@1.3.0~start: Returned: code: 1  signal: null
13 info lifecycle @elektra-web/client@1.3.0~start: Failed to exec start script
14 verbose stack Error: @elektra-web/client@1.3.0 start: `npm run build && cd ../webd && npm run build && npm run start:prod`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/lib/utils/lifecycle.js:255:16)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at EventEmitter.emit (events.js:191:7)
14 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/lib/utils/spawn.js:40:14)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at ChildProcess.emit (events.js:191:7)
14 verbose stack     at maybeClose (internal/child_process.js:920:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:230:5)
15 verbose pkgid @elektra-web/client@1.3.0
16 verbose cwd /home/markus/Projekte/Elektra/libelektra/src/tools/web/client
17 error Linux 4.12.0-0.bpo.2-amd64
18 error argv "/usr/bin/node" "/usr/bin/npm" "start"
19 error node v6.13.1
20 error npm  v3.10.10
21 error code ELIFECYCLE
22 error @elektra-web/client@1.3.0 start: `npm run build && cd ../webd && npm run build && npm run start:prod`
22 error Exit status 1
23 error Failed at the @elektra-web/client@1.3.0 start script 'npm run build && cd ../webd && npm run build && npm run start:prod'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the @elektra-web/client package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     npm run build && cd ../webd && npm run build && npm run start:prod
23 error You can get information on how to open an issue for this project with:
23 error     npm bugs @elektra-web/client
23 error Or if that isn't available, you can get their info via:
23 error     npm owner ls @elektra-web/client
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

But after that, webd suddenly works. Seems like something (docu and/or npm configs) are wrong.

But when connecting, I get immediately errors:

```
Command failed: "kdb" getmeta "user/metakey" "n\nh"
Metakey not found
```

Seems my like shell does not interprets `\n` like yours does.

When I try to add a key (which has a not-nice-looking hint) I get the error:

```
FetchError

invalid json response body at http://127.0.0.1:33333/kdb/system/test reason: Unexpected end of JSON input
```

Seems like you should test with Linux?

Please also finish all docu, so that I can make a code review.

omnidan commented:
@markus2330 I updated the api blueprints for both elektrad and webd and made sure all tests pass (via `npm test`). I also updated all README.md's and the diagrams. All tasks should be done now. Let me know what you think :grin:

omnidan commented:
Oops, I just realized that adjusting the API to fit the standards broke the client - fixing it.

markus2330 commented:
Please take your time and retest.

Some more dredd tests would not hurt (They ran through, despite the problems). CMake integration (so that dredd runs with `make run_all` would be very nice, too.

> reduces size of node_modules, the total build size of elektra-web is 288MB for me now

That is much better but far off where we would like to be for embedded systems. Btw. it seems like you already have the first customer (@tom-wa) but he also would prefer it to be smaller.

omnidan commented:
@markus2330 it seems like your issue was that webd is already running (maybe in a background process?). the other issues (invalid JSON response) had to do with the API updates (PUT, DELETE reuqests should not have a response body) and should be fixed now.

I also did a fresh clone again and followed the README.md (both running from source and cmake integration) to make sure everything works fine.

btw: documentation about the code structure is at the bottom of README.md.

> CMake integration (so that dredd runs with make run_all would be very nice, too.

I added the tests to the cmake file:

```
    add_test (NAME elektrad COMMAND npm test --unsafe-perm WORKING_DIRECTORY ${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad)
    add_test (NAME webd COMMAND npm test --unsafe-perm WORKING_DIRECTORY ${CMAKE_INSTALL_PREFIX}/${install_directory}/webd)
```

However, they do not seem to run when executing `make run_all` - how do I achieve this?

markus2330 commented:
> it seems like your issue was that webd is already running

I am pretty sure it did not run because I could start it afterwards (without stopping anything). Which line said that the port was taken?

> However, they do not seem to run when executing make run_all - how do I achieve this?

Yes, the add_test should do. Are you sure the cmake build folder is clean? Is ENABLE_TESTING enabled?

CMAKE_INSTALL_PREFIX, however, most likely is wrong. You cannot assume that it is installed when running tests.

omnidan commented:
> Which line said that the port was taken?

I saw this in your logs (before the npm output, which is unfortunately quite long and messy):

```
> @elektra-web/webd@1.3.0 start:prod /home/markus/Projekte/Elektra/libelektra/src/tools/web/webd
> cross-env NODE_ENV=production DEBUG=webd:*,-*debug,-*log node build/

  webd:info @elektra-web/webd v1.3.0 starting +0ms
  webd:info |- using default kdb command +43ms
  webd:info |- versions: { api: 1, elektra: { version: '0.8.23', major: 0, minor: 8, micro: 23 } } +1ms
  webd:client:info prod mode: serving client static assets +7ms
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE :::33334
```

---

> Yes, the add_test should do. Are you sure the cmake build folder is clean (and up-to-date)? Is ENABLE_TESTING enabled?

> CMAKE_INSTALL_PREFIX, however, most likely is wrong. You cannot assume that it is installed when running tests.

Thanks for the hints! It works now 😄 

```
        Start  32: elektra-web
 32/157 Test  #32: elektra-web ..............................   Passed   47.52 sec
```

markus2330 commented:
I still get the error

```
Command failed: "kdb" getmeta "user/metakey" "n\nh"
Metakey not found
```

though.

Please tell if everything (including docu) is fixed by setting ready to merge.

> Thanks for the hints!

Please also make sure to disable the tests if dredd is not available.

Btw. next to the huge hint in the "create new key", also the order of array keys which partly have subkeys (the one with subkeys are before the keys without subkeys) is irritating. I'll create new issues about that once we got this merged. I'll merge it if there are no major problems. These two problems are quite small, they should not stop us from conducting the user study. Obviously it is better if you can fix them.

omnidan commented:
@markus2330 I cannot reproduce this error... I will try setting up a debian VM. Are you using the real-world config or can you send me an export of the keys that cause this issue?

markus2330 commented:
Export of the metakey is, with ni:

```
;Ni1
; Generated by Nickel Plugin using Elektra (see libelektra.org).

 = 

[]
 a = b
 c = d
 e = f\\nk
 n = f\nh
 x = 
 n\nh = a
```


INI:

```
#@META a = b
#@META c = d
#@META e = f\nk
#@META n = f
h
#@META n
h = a
#@META x = 
[]
```

Maybe the literal \n, not an actual linebreak that causes the problem?

omnidan commented:
@markus2330 thank you! I could reproduce the issue with your export and my commit earlier fixes it!

markus2330 commented on doc/api_blueprints/elektrad.apib:
```diff
@@ -15,120 +15,156 @@ The API version is increased whenever breaking changes
 
 
-## elektra key database [/kdb/{+path}{?apikey}]
+## elektra key database [/kdb/{+path}]
```
I added the missing apikey to #1849

markus2330 commented on doc/api_blueprints/elektrad.apib:
```diff
@@ -15,120 +15,156 @@ The API version is increased whenever breaking changes
 ### get configuration [GET]
 
-+ Response 200 (application/json)
+this actually does `kdb get`, `kdb ls` and `kdb lsmeta`/`kdb getmeta` at once and is used to browse the kdb
```
why "actually"?, the rest of the sentence is also a rather poor description. ("at once"? "browse"?)

markus2330 commented on doc/api_blueprints/elektrad.apib:
```diff
@@ -15,120 +15,156 @@ The API version is increased whenever breaking changes
-        + value: hello world (string, required)
-        + meta (object, required) - metadata of the requested path
+        + exists: false (boolean, required)
+        + name: doesnotexist (string, required)
```
Why is the name added? (isn't this basename(path)?)

markus2330 commented on doc/api_blueprints/elektrad.apib:
```diff
@@ -15,120 +15,156 @@ The API version is increased whenever breaking changes
-+ Response 401 (application/json)
+### set configuration [PUT]
+
+works like `kdb set`
```
is there any difference?

markus2330 commented on doc/api_blueprints/elektrad.apib:
```diff
@@ -15,120 +15,156 @@ The API version is increased whenever breaking changes
++ Response 200
+
++ Response 400 (application/json; charset=utf-8)
     + Attributes (Error)
```
Which errors are possible?

markus2330 commented on doc/api_blueprints/elektrad.apib:
```diff
@@ -15,120 +15,156 @@ The API version is increased whenever breaking changes
-### set configuration [PUT]
 
-+ Request (text/plain)
+## metakeys [/kdbMeta/{+path}]
```
suddenly camelCase? Why kdbMeta additionally to kdb?

markus2330 commented on doc/api_blueprints/elektrad.apib:
```diff
@@ -15,120 +15,156 @@ The API version is increased whenever breaking changes
+
++ Response 204
+
++ Response 400 (application/json; charset=utf-8)
```
In general there should be much more here so that dredd has something to do.

errors? different requests?

Alternatively (or additionally) some curl test cases would be nice. (E.g. in shell recorder syntax.)

markus2330 commented on doc/api_blueprints/webd.apib:
```diff
@@ -80,122 +67,179 @@ for further information and explanation see [doc/VERSION.md](https://github.com/
+
++ Response 404
 
 ### set configuration [PUT]
```
Is this a copy&paste from elektrad or are there differences?

markus2330 commented on src/tools/web/client/package-lock.json:
```diff
@@ -1,6 +1,6 @@
 {
```
Is the file package-lock.json generated? How? What to do on security problems?

markus2330 commented on src/tools/web/client/package.json:
```diff
@@ -1,29 +1,32 @@
+  "repository": "github:ElektraInitiative/libelektra",
   "proxy": "http://localhost:33334",
-  "license": "BSD",
+  "license": "SEE LICENSE IN ../../../../LICENSE.md",
```
Isn't there a tag for our license? Isn't this tag parsed automatically?

markus2330 commented on src/tools/web/README.md:
```diff
@@ -78,30 +105,19 @@ Elektra web consists of multiple components:
 
- * [elektrad](https://master.libelektra.org/doc/api_blueprints/elektrad.apib), documentation: http://docs.elektrad.apiary.io/
- * [clusterd](https://master.libelektra.org/doc/api_blueprints/clusterd.apib), documentation: http://docs.clusterd.apiary.io/
+ * [elektrad](https://master.libelektra.org/doc/api_blueprints/elektrad.apib), documentation: https://elektrad.docs.apiary.io/
```
How to upload it there? It does not seem to be udpated?

markus2330 commented:
This is a huge improvement, thank you!

markus2330 commented:
I added some issues, mainly about visual appearance, in #1892

Please answer the review comments here or in #1892
